garage/envs/base.py

```diff
@@ -1,17 +1,156 @@
+"""Wrapper class that converts gym.Env into GarageEnv."""
 import collections
 
-from cached_property import cached_property
+import gym
+from gym.spaces import Box as GymBox
+from gym.spaces import Discrete as GymDiscrete
+from gym.spaces import Tuple as GymTuple
+
+from garage.core import Parameterized
+from garage.core import Serializable
+from garage.envs.env_spec import EnvSpec
+from garage.misc.overrides import overrides
+from garage.spaces import Box
+from garage.spaces import Discrete
+from garage.spaces import Product
+
+
+class GarageEnv(gym.Wrapper, Parameterized, Serializable):
+    """
+    Returns a Garage wrapper class for gym.Env.
+
+    In order to provide pickling (serialization) and parameterization
+    for gym.Envs, they must be wrapped with GarageEnv. This ensures
+    compatibility with existing samplers and checkpointing when the
+    envs are passed internally around garage.
+
+    Furthermore, GarageEnv silently converts action_space and
+    observation_space from gym.Spaces to garage.spaces.
+
+    Args:
+        env (gym.Env): the env that will be wrapped
+    """
+
+    def __init__(self, env):
+        Serializable.quick_init(self, locals())
+        Parameterized.__init__(self)
+        super(GarageEnv, self).__init__(env)
+
+    @overrides
+    def action_space(self):
+        """
+        Classes inheriting from GarageEnv need to convert
+        action_space from gym.space to garage.space.
+
+        Returns:
+            NotImplementedError
+        """
+        raise NotImplementedError
+
+    @overrides
+    def observation_space(self):
+        """
+        Classes inheriting from GarageEnv need to convert
+        observation_space from gym.space to garage.space.
+
+        Returns:
+            NotImplementedError
+        """
+        raise NotImplementedError
+
+    def close(self):
+        """
+        Close the wrapped env.
+
+        Returns:
+            None
+        """
+        self.env.close()
+
+    def get_params_internal(self, **tags):
```

Comment:
Perhaps raise a DeprecationWarning?

https://docs.python.org/3.6/library/warnings.html

Reply:
This will break existing functionality in the code

Reply:
Warnings print annoying error messages--they don't crash the code.

Reply:
It makes Travis fail, however:
```
======================================================================
ERROR: test_ddpg (algos.test_ddpg.TestDDPG)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/code/garage/tests/algos/test_ddpg.py", line 31, in test_ddpg
    algo.train()
  File "/root/code/garage/garage/theano/algos/ddpg.py", line 218, in train
    self.evaluate(epoch, pool)
  File "/root/code/garage/garage/theano/algos/ddpg.py", line 379, in evaluate
    self.env.log_diagnostics(paths)
  File "/root/code/garage/garage/envs/base.py", line 88, in log_diagnostics
    raise DeprecationWarning
DeprecationWarning
----------------------------------------------------------------------
Ran 54 tests in 783.391s
FAILED (errors=9)
```

Reply:
That's because you raised an exception, not a warning. In python you can raise anything as an exception, including the DeprecationWarning class.

https://pymotw.com/3/warnings/

Reply:
i see. thanks for the heads up

Reply:
👍